File: app/src/main/java/de/grobox/transportr/data/searches/SearchesRepository.java
package de.grobox.transportr.data.searches;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import de.grobox.transportr.AbstractManager;
import de.grobox.transportr.data.locations.FavoriteLocation;
import de.grobox.transportr.data.locations.LocationDao;
import de.grobox.transportr.favorites.trips.FavoriteTripItem;
import de.grobox.transportr.locations.WrapLocation;
import de.grobox.transportr.networks.TransportNetworkManager;
import de.schildbach.pte.NetworkId;

import static de.schildbach.pte.dto.LocationType.COORD;

public class SearchesRepository extends AbstractManager {

	private final SearchesDao searchesDao;
	private final LocationDao locationDao;

	private final LiveData<NetworkId> networkId;
	private final MediatorLiveData<List<FavoriteTripItem>> favoriteTripItems;

	@Inject
	public SearchesRepository(SearchesDao searchesDao, LocationDao locationDao, TransportNetworkManager transportNetworkManager) {
		this.searchesDao = searchesDao;
		this.locationDao = locationDao;
		this.networkId = transportNetworkManager.getNetworkId();
		LiveData<List<StoredSearch>> storedSearches = Transformations.switchMap(networkId, this::getStoredSearches);
		this.favoriteTripItems = new MediatorLiveData<>();
		this.favoriteTripItems.addSource(storedSearches, this::fetchFavoriteTrips);
	}

	private LiveData<List<StoredSearch>> getStoredSearches(NetworkId id) {
		return searchesDao.getStoredSearches(id);
	}

	private void fetchFavoriteTrips(List<StoredSearch> storedSearches) {
		runOnBackgroundThread(() -> {
			List<FavoriteTripItem> favoriteTrips = new ArrayList<>();
			for (StoredSearch storedSearch : storedSearches) {
				FavoriteLocation from = locationDao.getFavoriteLocation(storedSearch.fromId);
				FavoriteLocation via = storedSearch.viaId == null ? null : locationDao.getFavoriteLocation(storedSearch.viaId);
				FavoriteLocation to = locationDao.getFavoriteLocation(storedSearch.toId);
				if (from == null || to == null) throw new RuntimeException("Start or Destination was null");
				FavoriteTripItem item = new FavoriteTripItem(storedSearch, from, via, to);
				favoriteTrips.add(item);
			}
			favoriteTripItems.postValue(favoriteTrips);
		});
	}

	public LiveData<List<FavoriteTripItem>> getFavoriteTrips() {
		return favoriteTripItems;
	}

	public void storeSearch(FavoriteTripItem item) {
		if (item.getFrom().type == COORD || (item.getTo() != null && item.getTo().type == COORD)) return;
		runOnBackgroundThread(() -> {
			if (item.getUid() != 0) {
				searchesDao.updateStoredSearch(item.getUid(), new Date());
			} else {
				if (networkId.getValue() == null) return;
				NetworkId networkId = this.networkId.getValue();

				FavoriteLocation from = getFavoriteLocation(networkId, item.getFrom());
				FavoriteLocation via = getFavoriteLocation(networkId, item.getVia());
				FavoriteLocation to = getFavoriteLocation(networkId, item.getTo());
				if (from == null || to == null) throw new RuntimeException("Start or Destination wasn't saved");

				// try to find existing stored search
				StoredSearch storedSearch = searchesDao.getStoredSearch(networkId, from.getUid(), via == null ? null : via.getUid(), to.getUid());
				if (storedSearch == null) {
					// no search was found, so create a new one
					storedSearch = new StoredSearch(networkId, from, via, to);
				}
				searchesDao.storeSearch(storedSearch);
			}
		});
	}

	public void updateFavoriteState(FavoriteTripItem item) {
		if (item.getUid() == 0) throw new IllegalArgumentException();
		runOnBackgroundThread(() -> searchesDao.setFavorite(item.getUid(), item.isFavorite()));
	}

	public void removeSearch(FavoriteTripItem item) {
		if (item.getUid() == 0) throw new IllegalArgumentException();
		runOnBackgroundThread(() -> searchesDao.delete(item));
	}

	@Nullable
	private FavoriteLocation getFavoriteLocation(NetworkId networkId, @Nullable WrapLocation l) {
		if (l == null) return null;
		return locationDao.getFavoriteLocation(networkId, l.type, l.id, l.lat, l.lon, l.place, l.name);
	}

}

File: app/src/main/java/de/grobox/transportr/data/searches/SearchesRepository.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.data.searches

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.Transformations
import androidx.annotation.WorkerThread
import de.grobox.transportr.AbstractManager
import de.grobox.transportr.data.locations.FavoriteLocation
import de.grobox.transportr.data.locations.LocationDao
import de.grobox.transportr.favorites.trips.FavoriteTripItem
import de.grobox.transportr.networks.TransportNetworkManager
import de.schildbach.pte.NetworkId
import de.schildbach.pte.dto.LocationType.COORD
import java.util.*
import javax.inject.Inject

class SearchesRepository @Inject constructor(
        private val searchesDao: SearchesDao,
        private val locationDao: LocationDao,
        transportNetworkManager: TransportNetworkManager) : AbstractManager() {

    private val networkId: LiveData<NetworkId> = transportNetworkManager.networkId
    private val favoriteTripItems: MediatorLiveData<List<FavoriteTripItem>> = MediatorLiveData()

    val favoriteTrips: LiveData<List<FavoriteTripItem>>
        get() = favoriteTripItems

    init {
        val storedSearches = Transformations.switchMap<NetworkId, List<StoredSearch>>(networkId, searchesDao::getStoredSearches)
        this.favoriteTripItems.addSource(storedSearches, this::fetchFavoriteTrips)
    }

    private fun fetchFavoriteTrips(storedSearches: List<StoredSearch>?) {
        if (storedSearches == null) return
        runOnBackgroundThread {
            val favoriteTrips = ArrayList<FavoriteTripItem>()
            for (storedSearch in storedSearches) {
                val from = locationDao.getFavoriteLocation(storedSearch.fromId)
                val via = storedSearch.viaId?.let { locationDao.getFavoriteLocation(it) }
                val to = locationDao.getFavoriteLocation(storedSearch.toId)
                if (from == null || to == null) throw RuntimeException("Start or Destination was null")
                val item = FavoriteTripItem(storedSearch, from, via, to)
                favoriteTrips.add(item)
            }
            favoriteTripItems.postValue(favoriteTrips)
        }
    }

    @WorkerThread
    fun storeSearch(from: FavoriteLocation?, via: FavoriteLocation?, to: FavoriteLocation?): Long {
        if (from == null || to == null) return 0L
        if (from.type == COORD || via != null && via.type == COORD || to.type == COORD) throw IllegalStateException("COORD made it through")
        if (from.uid == 0L || to.uid == 0L) throw IllegalStateException("From or To wasn't saved properly :(")

        // try to find existing stored search
        var storedSearch = searchesDao.getStoredSearch(networkId.value!!, from.uid, via?.uid, to.uid)
        if (storedSearch == null) {
            // no search was found, so create a new one
            storedSearch = StoredSearch(networkId.value!!, from, via, to)
        }
        return searchesDao.storeSearch(storedSearch)
    }

    @WorkerThread
    fun isFavorite(uid: Long): Boolean {
        return searchesDao.isFavorite(uid)
    }

    fun updateFavoriteState(item: FavoriteTripItem) {
        updateFavoriteState(item.uid, item.isFavorite)
    }

    fun updateFavoriteState(uid: Long, isFavorite: Boolean) {
        if (uid == 0L) throw IllegalArgumentException()
        runOnBackgroundThread { searchesDao.setFavorite(uid, isFavorite) }
    }

    fun removeSearch(item: FavoriteTripItem) {
        if (item.getUid() == 0L) throw IllegalArgumentException()
        runOnBackgroundThread { searchesDao.delete(item) }
    }

}
